apps/system/js/audio_channel_controller.js

```diff
@@ -0,0 +1,211 @@
+      case 'activestatechanged':
+        var channel = evt.target;
+        // TODO: We should get the `isActive` state from evt.isActive.
+        // Then we don't need to do `channel.isActive()` here.
+        channel.isActive().onsuccess = (evt) => {
+          this._states.active = evt.target.result;
+          this.publish('statechanged');
+        };
+        break;
+
+      // There cannot be a `_terminated` event for a window element.
+      // Because the element doesn't exist
+      // when the `_terminated` event is triggered.
+      case '_terminating':
+        this.publish('terminated');
+        break;
+    }
+  };  
+
+  /**
+   * Set the policy for handling this audio channel.
+   *
+   * @param {Object} policy Policy to handle this audio channel.
+   * @param {Boolean} [policy.isAllowedToPlay] Play or pause it.
+   * @param {Boolean} [policy.isNeededToFadeOut] Fade out or not.
+   * @param {Boolean} [policy.isNeededToVibrate] Vibrate or not.
+   * @param {Boolean} [policy.isNeededToResumeWhenOtherEnds]
+   * Resume when other audio channel ends if as true.
+   * @return {AudioChannelController}
+   */
+  AudioChannelController.prototype.setPolicy = function(policy) {
+    this._policy = policy || this._policy;
+    return this;
+  };
+
+  /**
+   * Handle the audio channel with the policy,
+   * and clear the policy.
+   *
+   * @param {Function} callback Do something
+   * after the audio channel is handled by the policy.
+   */
+  AudioChannelController.prototype.proceedPolicy = function(callback) {
+    var policy = this._policy;
+    if (policy.isAllowedToPlay) {
+      this._play();
+    } else if (policy.isAllowedToPlay != null) {
+      this._pause();
+    }
+    if (policy.isNeededToFadeOut) {
+      this._fadeOut();
+    } else if (policy.isNeededToFadeOut != null) {
+      this._fadeIn();
+    }
+    if (policy.isNeededToVibrate) {
+      this._vibrate();
+    }
+    callback && callback(policy);
+    this._policy = {};
+  };
+
+  /**
+   * Get active state.
+   *
+   * @return {Boolean}
+   */
+  AudioChannelController.prototype.isActive = function() {
+    return this._states.active;
+  };
+
+  /**
+   * Get state of playing.
+   *
+   * @return {Boolean}
+   */
+  AudioChannelController.prototype.isPlaying = function() {
+    return this._states.playing;
+  };
+
+  /**
+   * Get state of fading out.
+   *
+   * @return {Boolean}
+   */
+  AudioChannelController.prototype.isFadingOut = function() {
+    return this._states.fadingOut;
+  };
+
+  /**
+   * Get state of vibrating.
+   *
+   * @return {Boolean}
+   */
+  AudioChannelController.prototype.isVibrating = function() {
+    return this._states.vibrating;
+  };
+
+  /**
+   * Play the audio channel.
+   */
+  AudioChannelController.prototype._play = function() {
+    var request = this._channel.setMuted(false);
+    this._states.playing = true;
+    request.onerror = function() {
+      this.app.debug('Cannot play the audio channel.');
+    };
```

Comment:
nit: Please use .then().catch() instead of onsuccess and onerror for DOM requests.